src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,219 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<E> {
+        private E item;
+        private Node<E> next;
+        private Node<E> prev;
+
+        Node(Node<E> prev, E element, Node<E> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<>(tail, value, null);
+        if (size == 0) {
+            addFirstElement(node);
+            return;
+        }
+        tail.next = node;
+        tail = node;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index);
+        Node<T> newNode = new Node<>(null, value, null);
+        Node<T> oldNode = getIndexNode(index);
+        if (size == 0) {
+            addFirstElement(newNode);
+            return;
+        }
+        if (index == size || index == 0) {
+            addTopBottom(newNode, index);
+            return;
+        }
+        newNode.prev = oldNode.prev;
+        newNode.next = oldNode;
+        oldNode.prev.next = newNode;
+        oldNode.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexGetSet(index);
+        Node<T> node = getIndexNode(index);
+        return node.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexGetSet(index);
+        Node<T> node = getIndexNode(index);
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexGetSet(index);
+        Node<T> node = getIndexNode(index);
+        T removedItem = node.item;
+        if (index == 0) {
+            removeBottom();
+            return removedItem;
+        }
+        if (index == size - 1) {
+            removeTop();
+            return removedItem;
+        }
+        removeNode(node);
+        return removedItem;
```

Comment:
the same is for method remove() is it really necessary to have different methods for these actions? Can we simplify this logic?